examples/vision/semantic_image_clustering.py

```diff
@@ -0,0 +1,577 @@
+from tqdm import tqdm
+
+"""
+## Prepare the data
+"""
+
+num_classes = 10
+input_shape = (32, 32, 3)
+
+(x_train, y_train), (x_test, y_test) = keras.datasets.cifar10.load_data()
+x_data = np.concatenate([x_train, x_test])
+y_data = np.concatenate([y_train, y_test])
+
+print("x_data shape:", x_data.shape, " - y_data shape:", y_data.shape)
+
+class_labels = [
+    "airplane",
+    "automobile",
+    "bird",
+    "cat",
+    "deer",
+    "dog",
+    "frog",
+    "horse",
+    "ship",
+    "truck",
+]
+
+"""
+## Define hyperparameters
+"""
+
+target_size = 32  # Resize the input images.
+representation_dim = 512  # The dimensions of the features vector.
+projection_units = 128  # The projection head of the representation learner.
+num_augumentations = 2  # Number of augmented images to generate for each input.
+num_clusters = 20  # Number of clusters.
+k_neighbours = 5  # Number of neighbors to consider during cluster learning.
+tune_encoder_during_clustering = False  # Freeze the encoder in the cluster learning.
+
+"""
+## Implement data preprocessing
+
+The data preprocessing step resizes the input images to the desired `target_size` and applies
+feature-wise normalization. Note that, when using `keras.applications.ResNet50V2` as the
+visual encoder, resizing the images into 255 x 255 inputs would lead to more accurate results 
+but require a longer time to train.
+"""
+
+data_preprocessing = keras.Sequential(
+    [
+        layers.experimental.preprocessing.Resizing(target_size, target_size),
+        layers.experimental.preprocessing.Normalization(),
+    ]
+)
+# Compute the mean and the variance from the data for normalization.
+data_preprocessing.layers[-1].adapt(x_data)
+
+"""
+## Data augmentation
+
+Unlink simCLR, which randomly picks a single data augmentation function to apply to an input
+image, we apply a set of data augmentation functions randomly to the input image.
+(You can experiment with other image augmentation techniques by following the [data augmentation tutorial](https://www.tensorflow.org/tutorials/images/data_augmentation).)
+"""
+
+data_augmentation = keras.Sequential(
+    [
+        layers.experimental.preprocessing.RandomTranslation(
+            (-0.2, 0.2), (-0.2, 0.2), fill_mode="nearest"
+        ),
+        layers.experimental.preprocessing.RandomFlip(mode="horizontal"),
+        layers.experimental.preprocessing.RandomRotation(
+            factor=0.15, fill_mode="nearest"
+        ),
+        layers.experimental.preprocessing.RandomZoom(
+            height_factor=(-0.3, 0.1), width_factor=(-0.3, 0.1), fill_mode="nearest"
+        ),
+        layers.experimental.preprocessing.RandomContrast(factor=0.4),
+        layers.Lambda(lambda images: tf.image.random_brightness(images, max_delta=0.4)),
+        layers.Lambda(
+            lambda images: tf.image.random_saturation(images, lower=0, upper=0.4)
+        ),
+        layers.Lambda(lambda images: tf.image.random_hue(images, max_delta=0.1)),
```

Comment:
These 3 transformations (brightness, saturation, hue) are largely useless because they're affine transformations. For instance "random brightness" just adds a random scalar value to your inputs, which isn't helpful in terms of data augmentation. You can skip them.

Reply:
Done. They are used by the original SCAN paper (but I am not sure how effective they were). But your comment makes sense.